karabo_data/reader.py

```diff
@@ -187,629 +128,195 @@ def __init__(self, prop, source):
-                index = pd.MultiIndex.from_arrays([trainids, pulse_id],
-                                                  names=['trainId', 'pulseId'])
-                # Does pulse-oriented data always have an extra dimension?
-                assert data.shape[1] == 1
-                data = data[:, 0]
-            data = data[:len(index)]
+        try:
+            return self._keys_cache[source]
+        except KeyError:
+            pass
+
+        if source in self.control_sources:
+            group = '/CONTROL/' + source
+        elif source in self.instrument_sources:
+            group = '/INSTRUMENT/' + source
         else:
-            raise ValueError("Unknown data source %r" % data_src)
+            raise SourceNameError(source)
 
-        return pd.Series(data, name=name, index=index)
+        res = set()
 
-    def get_dataframe(self, fields=(('*', '*'),), *, timestamps=False):
-        """Return a pandas dataframe for given data fields.
+        def add_key(key, value):
+            if isinstance(value, h5py.Dataset):
+                res.add(key.replace('/', '.'))
 
-        Parameters
-        ----------
-        fields : list of 2-tuples
-            Glob patterns to match device and field names, e.g.
-            ``("*_XGM/*", "*.i[xy]Pos")`` matches ixPos and iyPos from any XGM devices.
-            By default, all fields from all control devices are matched.
-        timestamps : bool
-            If false (the default), exclude the timestamps associated with each
-            control data field.
-        """
-        fields = _normalize_data_selection(fields, self)
-        if not timestamps:
-            fields = {(s, k) for (s, k) in fields if not k.endswith('.timestamp')}
-        control_series = [self.get_series(s, k) for (s, k) in fields]
+        self.file[group].visititems(add_key)
+        self._keys_cache[source] = res
+        return res
 
-        if not control_series:
-            return None
-        return pd.concat(control_series, axis=1)
+class DataCollection:
+    """An assemblage of data generated at European XFEL
 
-    def get_array(self, device, key, extra_dims=None):
-        """Return a labelled array for a particular data field.
+    Data consists of *sources* which each have *keys*. It is further
+    organised by *trains*, which are identified by train IDs.
 
-        The first axis of the returned data will be the train IDs.
-        Datasets which are per-pulse in the first dimension are not supported.
+    You normally get an instance of this class by calling :func:`H5File`
+    for a single file or :func:`RunDirectory` for a directory.
+    """
+    def __init__(self, files, selection=None, train_ids=None):
+        self.files = list(files)
 
-        Parameters
-        ----------
+        # selection: {source: set(keys)}
+        # None as value -> all keys for this source
+        if selection is None:
+            selection = {}
+            for f in self.files:
```

Comment:
We should figure out what to do with name collisions. It's more complex than just the same name in multiple files, though, because we expect the sequence files to have the same sources. It should probably be files containing the same source with the same train IDs.

I suspect the only logical thing to do is throw an error in that case - I don't see any way we can figure out which device to prefer.